Document TokenProvider class (#7475)
microsoft/FluidFramework: docs/content/docs/build/auth.md

### A secure serverless token provider

One option for building a secure token provider is to create a serverless Azure Function and expose it. This enables you to store the *tenant secret key* on a secure server. Your application calls the Azure Function to generate tokens rather than signing them locally like the `InsecureTokenProvider` does.

> You can create the token provider class say `AzureFunctionTokenProvider` which extends the `ITokenProvider` interface without exposing the tenant key secret in client-side code. This class would be responsible for fetching the token from your very own backend. It accepts the url to Azure function and an optional user object.

Reviewer: I would rephrase the "say `AzureFunctionTokenProvider`" part like this:

You can create a token provider class which extends the `ITokenProvider` interface without exposing the tenant key secret in client-side code. In the example below the token provider class is called `AzureFunctionTokenProvider`.

```typescript
import { ITokenProvider, ITokenResponse } from "@fluidframework/azure-client";

export class AzureFunctionTokenProvider implements ITokenProvider {
  constructor(
    private readonly azFunctionUrl: string,

>     private readonly user?: Pick<AzureMember, "userId" | "userName" | "additionalDetails">,

Reviewer: Is there a simpler way to express this with a little less "type magic?" No worries if not.

Reviewer 2: For brevity we can just use AzureMember here. :)

```suggestion
    private readonly user?: AzureMember,
```

  );

  fetchOrdererToken(tenantId: string, documentId: string, refresh?: boolean): Promise<ITokenResponse>;

>   fetchStorageToken(tenantId: string, documentId: string, refresh?: boolean): Promise<ITokenResponse>;

Reviewer: Do these intentionally omit the implementation? If so, I suggest making the code reflect that, and add a comment like `// omitted` so it's clear.

}
```

> `fetchOrdererToken` and `fetchStorageToken` are responsible for fetching the orderer and storgae url from the host respectively. They return `TokenResponse` object representing token value along with flag indicating whether token came from cache.

Reviewer 3: ```suggestion
`fetchOrdererToken` and `fetchStorageToken` are responsible for fetching the orderer and storage url from the host respectively. They return `TokenResponse` object representing token value along with flag indicating whether token came from cache.
```

> `fetchOrdererToken` and `fetchStorageToken` are responsible for fetching the orderer and storgae url from the host respectively. They return `TokenResponse` object representing token value along with flag indicating whether token came from cache.

Reviewer: ```suggestion
The `fetchOrdererToken` and `fetchStorageToken` functions are responsible for fetching the orderer and storage URLs from the host. Both functions return `TokenResponse` objects representing the token value along with a flag indicating whether token came from the cache.
```

Reviewer: I think this is the first time you mention the cache. Is it important to note that here? If not, just omit it.

> To ensure that the secret doesn't get exposed, it is passed to a secure, backend Azure function from which the token is fetched. One of the possible way to fetch the token is making an axios `GET` request call to your Azure function by passing in the tenantID, documentId and userID/userName as optional parameters. Azure function is responsible for mapping between the tenant ID to a tenant key secret to generate and sign the token such that the service will accept it.

Reviewer: ```suggestion
To ensure that the tenant secret key is kept secure, it is stored in a secure backend location and is only accessible from within the Azure Function. One way to fetch a signed token is to make a `GET` request to your Azure Function, providing the `tenantID` and `documentId`, and optionally `userID`/`userName`. The Azure Function is responsible for the mapping between the tenant ID and a tenant key secret to appropriately generate and sign the token such that the Azure Fluid Relay service will accept it.
```

```typescript
private async getToken(tenantId: string, documentId: string): Promise<string> {

>   return axios.get(this.azFunctionUrl, {

Reviewer: I don't have any problem with axios, but is there a vanilla JS way to do this that isn't ugly or illegible? If not, then I suggest adding a callout above this code sample that says, "The example below uses the axios (add a link to the npm package) library to make HTTP requests. You can use other libraries or approaches to making an HTTP request."

Reviewer 2: ^^ This is why my suggestion was to have this as a `Recipe` or something talking about how to build a specific scenario. Then we have a more basic doc that talks about the principals where you can sudo code something like:

```js
private async getToken(tenantId, documentId) {
  // make call to backend service to retrieve server generated token 
  const token = getTokenFromServer(...);
  return token;
```

      params: {
          tenantId,
          documentId,
          userId: this.user?.userId,
          userName: this.user?.userName,
          additionalDetails: this.user?.additionalDetails,
      },
    }).then((response) => {
        return response.data as string;
    }).catch((err) => {
        return err as string;
  });
}
```

> 

Reviewer 4: Let's also consider including some documentation here for what is happening on the server-side with `generateToken`, i.e. within the Azure function. This is so that if the developers choose to use some implementation other than an Azure Function, they will have some directions on how to set that up.

## Adding custom data to tokens

> You can add custom data such as email id, gender, address, etc for your token generation. See [Connect to an Azure Fluid Relay service]({{< relref "azure-frs.md" >}}) for more information.

Reviewer 4: I think since we are also discussing `additionalDetails` in the code snippet above, we can include a line that says that these values can be added in there
